src/icu_win.cc

```diff
@@ -0,0 +1,242 @@
+// Copyright 2025 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   https://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+#include "icu_win.h"
+
+#if defined(_WIN32)
+
+#if !defined(NOMINMAX)
+#define NOMINMAX
+#endif  // !defined(NOMINMAX)
+#include <windows.h>
+
+#include <atomic>
+#include <cstdint>
+#include <limits>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+namespace cctz {
+namespace icu {
+namespace {
+
+// UChar is defined as char16_t in ICU by default, but it is also safe to assume
+// wchar_t and char16_t are equivalent on Windows.
+using UChar = wchar_t;
+
+enum UErrorCode : std::int32_t {
+  U_ZERO_ERROR = 0,
+  U_BUFFER_OVERFLOW_ERROR = 15,
+};
+
+bool U_SUCCESS(UErrorCode error) { return error <= U_ZERO_ERROR; }
+
+// ICU function signatures
+using ucal_getHostTimeZone_func = std::int32_t(__cdecl*)(
+    UChar* result, std::int32_t result_capacity, UErrorCode* status);
+using ucal_getTimeZoneIDForWindowsID_func = std::int32_t(__cdecl*)(
+    const UChar* winid, std::int32_t len, const char* region, UChar* id,
+    std::int32_t id_capacity, UErrorCode* status);
+
+std::atomic<bool> g_unavailable;
+std::atomic<ucal_getHostTimeZone_func> g_ucal_getHostTimeZone;
+std::atomic<ucal_getTimeZoneIDForWindowsID_func>
+    g_ucal_getTimeZoneIDForWindowsID;
+
+struct IcuFunctions {
+  const bool available;
+  const ucal_getHostTimeZone_func ucal_getHostTimeZone;
+  const ucal_getTimeZoneIDForWindowsID_func ucal_getTimeZoneIDForWindowsID;
+};
+
+IcuFunctions Unavailable() { return {false, nullptr, nullptr}; }
+
+template <typename T>
+static T AsProcAddress(HMODULE module, const char* name) {
+  static_assert(
+      std::is_pointer<T>::value &&
+          std::is_function<typename std::remove_pointer<T>::type>::value,
+      "T must be a function pointer type");
+  const auto proc_address = ::GetProcAddress(module, name);
+  return reinterpret_cast<T>(reinterpret_cast<void*>(proc_address));
+}
+
+std::wstring GetSystem32Dir() {
+  size_t dynamic_buffer_size = 0;
+  {
+    // Fast-path. Usually it's "C:\Windows\System32", which is 19 characters.
+    const size_t buffer_size = 32;
+    wchar_t buffer[buffer_size];
```

Comment:
Prefer to use std::array for these buffers here and in the rest of the file. With hardened standard libraries being deployed, bounds checking is possible (I don't see much benefit in this code, but it is a good habit.) You also get to use buffer.size() as the constant size bound to the buffer.

Reply:
Done.